pkg/private/pkg_files.bzl

```diff
@@ -0,0 +1,305 @@
+                    # the important part. The likely behavior people want is
+                    # just the content, so we strip the directory name.
+                    dest = '/'.join(d_path.split('/')[0:-1])
+                    add_tree_artifact(content_map, dest, f, src.label)
+                else:
+                    add_single_file(content_map, d_path, f, src.label)
+
+def add_single_file(content_map, dest_path, src, origin, mode=None, user=None, group=None):
+    """Add an single file to the content map.
+
+    Args:
+      content_map: The content map
+      dest_path: Where to place the file in the package.
+      src: Source object. Must have len(src[DefaultInfo].files) == 1
+      origin: The rule instance adding this entry
+      mode: fallback mode to use for Package*Info elements without mode
+      user: fallback user to use for Package*Info elements without user
+      group: fallback mode to use for Package*Info elements without group
+    """
+    dest = dest_path.strip('/')
+    _check_dest(content_map, dest, origin)
+    content_map[dest] = _DestFile(
+        src = src,
+        origin = origin,
+        mode = mode,
+        user = user,
+        group = group,
+    )
+
+def add_tree_artifact(content_map, dest_path, src, origin, mode=None, user=None, group=None):
+    """Add an single file to the content map.
+
+    Args:
+      content_map: The content map
+      dest_path: Where to place the file in the package.
+      src: Source object. Must have len(src[DefaultInfo].files) == 1
+      origin: The rule instance adding this entry
+      mode: fallback mode to use for Package*Info elements without mode
+      user: fallback user to use for Package*Info elements without user
+      group: fallback mode to use for Package*Info elements without group
+    """
+    content_map[dest_path] = _DestFile(
+        src = src,
+        origin = origin,
+        is_dir = 2,
+        mode = mode,
+        user = user,
+        group = group,
+    )
+
+def write_manifest(ctx, manifest_file, content_map):
+    """Write a content map to a manifest file.
+
+    The format of this file is currently undocumented, as it is a private
+    contract between the rule implementation and the package writers.  It will
+    become a published interface in a future release.
+
+    For reproducibility, the manifest file must be ordered consistently.
+    """
+    ctx.actions.write(
+        manifest_file,
+        "[\n" + ",\n".join(
+            [_encode_manifest_entry(dst, content_map[dst])
+             for dst in sorted(content_map.keys())]
+            ) + "\n]\n"
+    )
+
+def _encode_manifest_entry(dest, df):
+    return json.encode([
+        dest.strip('/'),
+        df.src.path if df.src else None,
+        df.mode or "",
+        df.user or None,
+        df.group or None,
+        df.link_to if hasattr(df, "link_to") else "",
+        df.is_dir if hasattr(df, "is_dir") else 0,
```

Comment:
Shouldn't `origin` be encoded here?  Collisions between TreeArtifacts and normal source files can't be consistently detected in Starlark, and origins might be useful information to have in the tar builder.

Reply:
Maybe. I'm not sure. For now, I don't want to increase the file size or take the time to create the code/tests to look for a hypothetical problem.

Reply:
Fair enough.  We haven't directly encountered this problem ourselves, FWIW.  It just occurred to me as I was working on the analogue for `pkg_rpm`.